tutorial on fdp support in xNVMe (#357)
OpenMPDK/xNVMe: docs/tutorial/fdp/index.rst

.. _sec-tutorials-fdp-intro:

=========================
 Flexible Data Placement
=========================

> In this section, you will find the guide on **FDP** aka Flexible data
> placement support in xNVMe.

Reviewer: ```suggestion
In this section, you will find a guide on **FDP** aka Flexible Data
Placement support in xNVMe.
```

This will guide you with examples of all the supported **FDP** log pages, set
get features, I/O management commands. This will also cover write command with

> hints from fio xnvme ioengine perspective.

Reviewer: ```suggestion
hints from the perspective of FIO's xNVMe ioengine.
```

Concepts and Prelude
--------------------

Flexible data placement adds an enhancement to NVM Command Set by enabling
host guided data placement. This introduces a Reclaim Unit (**RU**) which is a

> logical representation of non-volatile storage within a reclaim group that is

Reviewer: ```suggestion
logical representation of non-volatile storage within a Reclaim Group that is
```

able to be physically erased by the controller without disturbing any other

> reclaim units.

Reviewer: ```suggestion
Reclaim Units.
```

> Placement Identifier is a data structure that specifies a reclaim group
> identifier and a placement handle that references a reclaim unit.

Reviewer: ```suggestion
Placement Identifier is a data structure that specifies a Reclaim Group
Identifier and a placement handle that references a Reclaim Unit.
```

> Placement Handle is a namespace scoped handle that maps to an endurance group
> scoped reclaim unit handle which references a reclaim unit in each reclaim
> group.

Reviewer: ```suggestion
Placement Handle is a namespace scoped handle that maps to an Endurance Group
scoped Reclaim Unit Handle which references a Reclaim Unit in each Reclaim
Group.
```

Reclaim Unit Handle (**RUH**) is a controller resource that references a

> reclaim unit in each reclaim group.

Reviewer: ```suggestion
Reclaim Unit in each Reclaim Group.
```

For the complete information on FDP, please see the ratified technical proposal
TP4146 Flexible Data Placement 2022.11.30 Ratified, which can be found here
https://nvmexpress.org/wp-content/uploads/NVM-Express-2.0-Ratified-TPs_20230111.zip

Get log page
------------

> There are 4 new log pages associated with **FDP**. These are FDP configuration,
> reclaim unit handle usage, FDP statistics and FDP events. All these log pages

Reviewer: ```suggestion
There are 4 new log pages associated with **FDP**. These are FDP Configuration,
Reclaim Unit Handle Usage, FDP Statistics and FDP Events. All these log pages
```

> are endurance group scoped and hence you need to specify endurance group
> identifier in log specific identifier field.

Reviewer: ```suggestion
are Endurance Group scoped and hence you need to specify an Endurance Group
Identifier in the Log Specific Identifier field.
```

For all the above 4 mentioned log page, you can use **xNVMe** CLI:

> The fdp configuration log page requires dynamic memory allocation, as there

Reviewer: ```suggestion
The FDP Configuration log page requires dynamic memory allocation, as there
```

> can be multiple configuration, with each having multiple reclaim unit handles.

Reviewer: ```suggestion
can be multiple configurations each having multiple Reclaim Unit Handles.
```

You will have to specify the data size in bytes. The command can be run like:

.. literalinclude:: 100_xnvme_log_fdp_config.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 100_xnvme_log_fdp_config.out
   :language: bash

> For fdp statistics log page the command can be run like:

Reviewer: ```suggestion
For the FDP Statistics log page the command can be run like:
```

.. literalinclude:: 200_xnvme_log_fdp_stats.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 200_xnvme_log_fdp_stats.out
   :language: bash

> Similar to fdp configuration log page you will have to specify the number of

Reviewer: ```suggestion
Similar to the FDP Configuration log page, you will have to specify the number of
```

> recalim unit handle usage descriptors to fetch. The command can be run like:

Reviewer: ```suggestion
Reclaim Unit Handle Usage descriptors to fetch. The command can be run like:
```

.. literalinclude:: 300_xnvme_log_ruhu.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 300_xnvme_log_ruhu.out
   :language: bash

> The fdp events log page will have multiple events. You will have to specify

Reviewer: ```suggestion
The FDP Events log page will have multiple events. You will have to specify
```

> number of events you want to fetch. You also need to specify whether you need

Reviewer: ```suggestion
the number of events you want to fetch. You also need to specify whether you need
```

host or controller events. This can be done by log specific parameter. The
complete command can be run like:

.. literalinclude:: 400_xnvme_log_fdp_events.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 400_xnvme_log_fdp_events.out
   :language: bash

Set and get-feature
-------------------

> There are 2 new set and get-feature commands, introduced with FDP. These are
> Flexible data placement which controls operation of FDP capability in the
> specified endurance group, FDP events which controls if a controller generates
> FDP events associated with a specific reclaim unit handle.

Reviewer: ```suggestion
There are 2 new Set and Get Feature commands, introduced with **FDP**. These are
Flexible Data Placement, which controls operation of FDP capability in the
specified endurance group, and FDP Events, which controls if a controller generates
FDP Events associated with a specific Reclaim Unit Handle.
```

> xNVMe does not support namespace management commands. Thus we cannot enable or
> disable FDP by sending set-feature command to the endurance group, as it
> requires deletion of all namespaces in that endurance group. However you can
> check the FDP capability by sending a get-feature command. The command can be

Reviewer: ```suggestion
xNVMe does not support Namespace Management commands. Thus we cannot enable or
disable **FDP** by sending Set Feature command to the Endurance Group, as it
requires deletion of all namespaces in that Endurance Group. However you can
check the FDP capability by sending a Get Feature command. The command can be
```

run like:

.. literalinclude:: 010_xnvme_feature_get.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 010_xnvme_feature_get.out
   :language: bash

> Command dword 12 controls whether you want to enable or disable FDP events.

Reviewer: ```suggestion
Command Dword 12 controls whether you want to enable or disable FDP Events.
```

> You will have to specify number of events to enable or disable and placement
> handle associated with it. These will be part of feat field. This will require
> you to specify the size in bytes of data buffer. To enable all the fdp events

Reviewer: ```suggestion
You will have to specify number of events to enable or disable and the Placement
Handle associated with it. These will be part of the Feat field. This will require
you to specify the size in bytes of the data buffer. To enable all the FDP Events
```

you can run command like:

.. literalinclude:: 020_xnvme_set_fdp_events.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 020_xnvme_set_fdp_events.out
   :language: bash

> You can get the status of supported fdp events. The command dowrd 11 remains
> same as the set-feature command. You can run the command like:

Reviewer: ```suggestion
You can get the status of supported FDP Events. The Command Dword 11 remains the
same as the Set Feature command. You can run the command like:
```

.. literalinclude:: 030_xnvme_feature_get.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 030_xnvme_feature_get.out
   :language: bash

> I/O management

Reviewer: ```suggestion
I/O Management
```

--------------

> Two I/O management commands are introduced with FDP. These are I/O management
> send and I/O management receive.

Reviewer: ```suggestion
Two I/O Management commands are introduced with **FDP**. These are I/O Management
Send and I/O Management Receive.
```

> I/O management receive supports reclaim unit handle status command. You will
> have to specify the number of recalim unit handle status descriptors to fetch.

Reviewer: ```suggestion
I/O Management Receive supports the Reclaim Unit Handle Status command. You will
have to specify the number of Reclaim Unit Handle Status descriptors to fetch.
```

You can run the command like:

.. literalinclude:: 040_xnvme_fdp_ruhs.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 040_xnvme_fdp_ruhs.out
   :language: bash

> I/O management send supports reclaim unit handle update command. You will have
> to specify placement identifier for this. You can run the command like:

Reviewer: ```suggestion
I/O Management Send supports Reclaim Unit Handle Update command. You will have
to specify a Placement Identifier for this. You can run the command like:
```

.. literalinclude:: 050_xnvme_fdp_ruhu.cmd
   :language: bash

The command should produce output similar to:

.. literalinclude:: 050_xnvme_fdp_ruhu.out
   :language: bash

xnvme ioengine
--------------

> FIO's xnvme ioengine provides fdp support since 3.35 release.
> For more information on the fdp specific options you can look at:

Reviewer: ```suggestion
FIO's xNVMe ioengine provides FDP support since the 3.35 release.
For more information on the FDP specific options you can look at:
```
